### PR TITLE
Replace NI Advanced call with synchronous version

### LINKED ISSUE #64: Add Insights Advanced sync callback
Necessary for nexmo/nexmo-cli#87

### PR DESCRIPTION
**Pull request by cbetta**
- Moved async call into own option
- Fixed all API endpoints to ones used in current docs
- Updated documentation to highlight changes

This will probably require a minor release bump as this commit will break for anyone expecting `nexmo.numberInsight.get({level: 'advanced'})` to be asynchronous.

Closes #64

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -13,8 +13,9 @@ var checkVerifyEndpoint = {host:'api.nexmo.com',path:'/verify/check/json'};
 var controlVerifyEndpoint = {host:'api.nexmo.com',path:'/verify/control/json'};
 var searchVerifyEndpoint = {host:'api.nexmo.com',path:'/verify/search/json'};
 var niEndpoint = {host:'rest.nexmo.com',path:'/ni/json'};
-var niBasicEndpoint = {host:'api.nexmo.com',path:'/number/format/json'};
-var niStandardEndpoint = {host:'api.nexmo.com',path:'/number/lookup/json'};
+var niBasicEndpoint = {host:'api.nexmo.com',path:'/ni/advanced/async/json'};
```
Comment (leggetter) [inline src/index.js]: The variable name and `path` value don't seem to match up.

File: src/NumberInsight.js [hunk 2]
```diff
@@ -33,55 +33,59 @@ class NumberInsight {
...
     delete options.level;
-    
+
     if(level === 'advanced') {
+      this._nexmo.numberInsightAdvanced.apply(this._nexmo, arguments);
+    } else if (level === 'asyncAdvanced') {
       this._nexmo.numberInsight.apply(this._nexmo, arguments);
     }
     else if(level === 'standard') {
+      console.log('3');
```
Comment (leggetter) [inline src/NumberInsight.js]: Debug left in.

File: src/NumberInsight.js [hunk 3]
```diff
@@ -33,55 +33,59 @@ class NumberInsight {
...
       this._nexmo.numberInsightStandard.apply(this._nexmo, arguments);
     }
     else {
+      console.log('4');
```
Comment (leggetter) [inline src/NumberInsight.js]: Debug left in.

File: README.md [hunk 4]
```diff
@@ -347,6 +349,20 @@ nexmo.number.buy(countryCode, msisdn, callback);
 ```
 
 ### Cancel Number
+=======
```
Comment (leggetter) [inline README.md]: Is this required? The `###` should be enough and I think all other headers use this syntax.

File: README.md [hunk 5]
```diff
@@ -333,6 +333,8 @@ nexmo.number.search(countryCode,options,callback);
 	* `size`
 
 For more details on what the above options mean refer to the Nexmo API  [documentation](https://docs.nexmo.com/tools/developer-api/number-search)
+=======
```
Comment (leggetter) [inline README.md]: Required?

File: README.md (hunk 4 above)
Comment (leggetter) [reply README.md]: I'm not sure why this seems to be followed by code. Maybe the GitHub preview has gone wonky?

I see:

``````
### Cancel Number
======

nexmo.numberInsight.get({level: 'advanced', number: NUMBER}, consolelog);
\```

For more info...
``````

(ignore the backslash)

**Review by leggetter — CHANGES_REQUESTED** [review]
There seem to be quite a few whitespaces changes in the code. Not sure why that is, and I'm not going to get to worried about it :)

Main thing is that the README seems to have been broken. I think a bunch of text may have accidentally been removed.

File: README.md (hunk 5 above)
Comment (leggetter) [reply README.md]: I think this is where a mistake has been made and text accidentally removed.

**Comment by cbetta** [comment]
@leggetter Damn, that's not right. Seems my rebase didn't go well. I'll have a look

File: README.md [hunk 6]
```diff
@@ -210,23 +210,33 @@ nexmo.numberInsight.get({level: 'basic', number: '1-234-567-8900'}, consolelog);
...
 ```js
-nexmo.numberInsight.get({level: 'standard', number: '1-234-567-8900'}, consolelog);
+nexmo.numberInsight.get({level: 'standard', number: '1-234-567-8900'}, callback_function);
 ```
 
 ### Advanced
 
 ```js
-nexmo.numberInsight.get({level: 'advanced', number: NUMBER}, callback);
+nexmo.numberInsight.get({level: 'advanced', number: NUMBER}, callback_function);
```
Comment (leggetter) [inline README.md]: `callback_function` -> `callback` - Looks like your rubyism has snuck in!

File: README.md [hunk 7]
```diff
@@ -210,23 +210,33 @@ nexmo.numberInsight.get({level: 'basic', number: '1-234-567-8900'}, consolelog);
...
 ```
 
 For more information check the documentation at https://docs.nexmo.com/number-insight/advanced
 
+### Advanced Async
+
+Number Insight Advanced might take a few seconds to return a result, therefore the option exist to process the result asynchronously through a webhook.
+
+```js
+nexmo.numberInsight.get({level: 'advancedAsync', number: NUMBER, callback: "http://example.com"}, callback_function);
```
Comment (leggetter) [inline README.md]: just `callback` is cool

**Review by leggetter — COMMENTED** [review]
I'd suggest we stick with the consistent snakeCase naming convention of variables. I can live with the whitespace changes :)

File: README.md (hunk 6 above)
Comment (cbetta) [reply README.md]: @leggetter haha, sorry. Was just trying to clarify here there's a difference in the Async call between the callback URL for the webhook and the callback function for this function.

Want me to change it to `callbackFunction`?

Additionally we should add this to the linter to make sure we ensure camelCase.

File: README.md (hunk 7 above)
Comment (cbetta) [reply README.md]: See other comment

File: README.md (hunk 6 above)
Comment (leggetter) [reply README.md]: nah - just use `callback`. All/most of the other examples just use that.

File: README.md (hunk 5 above)
Comment (cbetta) [reply README.md]: Renamed it to `callback` everywhere

File: README.md (hunk 6 above)
Comment (cbetta) [reply README.md]: @leggetter fixed. And updated all other mentions to callback too instead of `consolelog` which seems to not be used anywhere else.

**Review by leggetter — APPROVED** [review]

**Comment by leggetter** [comment]
@cbetta - so, now a version bump and the need to publish?

@tjlytle - worth your input on the decision we make for the version. There's a breaking change here with one of the APIs since insight advanced used to be async. It's now sync. There's a specific async version now.

**Comment by tjlytle** [comment]
@leggetter @cbetta Any way to keep the existing behaviour in _this_ version and change in a future major release? `advancedSync` would be the new `level`.

**Comment by cbetta** [comment]
@leggetter @tjlytle I'd say bump it to `1.1.0`. This is what semver is for.

**Comment by leggetter** [comment]
If it's a breaking API change then in theory we should be changing the MAJOR. So that's 2.0.0.

**Comment by cbetta** [comment]
Ah I stand corrected. Yeah so technically 2.0.0. I'm happy with that. Embrace Chrome's "don't be afraid to increase the major" motto!

Also happy to make the change minor ;)

**Comment by tjlytle** [comment]
I'd rather save 2.0 for a more significant change of the library (which
will likely be sooner than later).

**Comment by cbetta** [comment]
@tjlytle although I understand your intent I also want to point out that that's a completely cosmetic decision. Version numbers are meant to signify change in a semantic way, not to celebrate some milestone in the code base. 

Anyway, don't want to be a blocker so I'm happy to change the API to `advanced` and `advancedSync` to keep the code backwards compatible. @leggetter thoughts?

**Pull request merged**